src/main/scala/firrtl/transforms/GroupComponents.scala

```diff
@@ -0,0 +1,340 @@
+
+            // Add connection in Top to group's Module port
+            added += Connect(NoInfo, WSubField(WRef(label2instance(group)), toPort), otherExp)
+
+            // Return WRef with new kind (its inside the group Module now)
+            WRef(toPort, otherExp.tpe, PortKind, MALE)
+
+          // case 3: source in different group
+          case otherGroup =>
+            // Add port to otherGroup's Module
+            val fromPort = punchSignalOut(otherGroup, otherExp)
+            val toPort = addPort(group, otherExp, Input)
+
+            // Add connection in Top from otherGroup's port to group's port
+            val groupInst = label2instance(group)
+            val otherInst = label2instance(otherGroup)
+            added += Connect(NoInfo, WSubField(WRef(groupInst), toPort), WSubField(WRef(otherInst), fromPort))
+
+            // Return WRef with new kind (its inside the group Module now)
+            WRef(toPort, otherExp.tpe, PortKind, MALE)
+        }
+    }
+
+    // Given the sink is in the parent module, tidy up source references belonging to groups
+    def inTopFixExps(e: Expression): Expression = e match {
+      case _: DoPrim|_: Mux|_: ValidIf => e map inTopFixExps
+      case otherExp: Expression =>
+        val wref = getWRef(otherExp)
+        if(byNode(wref.name) != "") {
+          // Get the name of source's group
+          val otherGroup = byNode(wref.name)
+
+          // Add port to otherGroup's Module
+          val otherPortName = punchSignalOut(otherGroup, otherExp)
+
+          // Return WSubField (its inside the top Module still)
+          WSubField(WRef(label2instance(otherGroup)), otherPortName)
+
+        } else otherExp
+    }
+
+    def onStmt(s: Statement): Statement = {
+      s match {
+        // Sink is in a group
+        case r: IsDeclaration if byNode(r.name) != "" =>
+          val topStmts = mutable.ArrayBuffer[Statement]()
+          val group = byNode(r.name)
+          groupStatements(group) += r mapExpr inGroupFixExps(group, topStmts)
+          Block(topStmts)
+        case c: Connect if byNode(getWRef(c.loc).name) != "" =>
+          // Sink is in a group
+          val topStmts = mutable.ArrayBuffer[Statement]()
+          val group = byNode(getWRef(c.loc).name)
+          groupStatements(group) += Connect(c.info, c.loc, inGroupFixExps(group, topStmts)(c.expr))
+          Block(topStmts)
+        // TODO Attach if all are in a group?
+        case _: IsDeclaration|_: Connect|_: Attach =>
+          // Sink is in Top
+          val ret = s mapExpr inTopFixExps
+          ret
+        case other => other map onStmt
+      }
+    }
+
+
+    // Build datastructures
+    val newTopBody = Block(labelOrder.map(g => WDefInstance(NoInfo, label2instance(g), label2module(g), UnknownType)) ++ Seq(onStmt(m.body)))
+    val finalTopBody = Block(Utils.squashEmpty(newTopBody).asInstanceOf[Block].stmts.distinct)
+
+    val newModules = labelOrder.filter(_ != "") map { group =>
```

Comment:
Why is this filtering for empty labels? What does it mean if they're empty?

Reply:
The empty label indicates the original module's group. This means that it won't be grouped with any of the specified groups, but will remain in the original module.

Reply:
I will add some comments to clarify.